ql/ql/src/codeql_ql/style/RedundatAggregateQuery.qll

```diff
@@ -0,0 +1,50 @@
+import ql
+
+/**
+ * Holds if `aggr` is of one of the following forms:
+ * `exists(var | range | formula)` or `any(var | range | formula)`
+ */
+private predicate aggregate(AstNode aggr, Formula range, AstNode formula, VarDecl var) {
+  exists(Exists ex | aggr = ex |
+    ex.getRange() = range and
+    ex.getFormula() = formula and
+    count(ex.getArgument(_)) = 1 and
+    ex.getArgument(0) = var
+  )
+  or
+  exists(Any anyy | aggr = anyy |
+    anyy.getRange() = range and
+    anyy.getExpr(0) = formula and
+    count(anyy.getExpr(_)) = 1 and
+    count(anyy.getArgument(_)) = 1 and
+    anyy.getArgument(0) = var
+  )
+}
+
+/**
+ * Holds if `aggr` is a redundant aggregate.
+ * The aggregate declares a single variable `var`, the value of which is `operand`, and it is only used in `formula`.
+ */
+predicate redundatAggregate(AstNode aggr, AstNode formula, VarDecl var, AstNode operand) {
```

Comment:
Random question, would this also consider rewriting aggregates as inline casts, e.g.
```
exists(Type t | t = foo())
```
into
```
foo().(Type)
```

Reply:
It doesn't currently, but it could.  

Your example would become `foo() instanceof Type` as `exists(..)` is a formula and not an expression. 

(`any(Type t | t = foo())` would become `foo().(Type)`).

Reply:
And I've done it. 
The patch is grouped together one. 
But if you check out the `run the redundant-inline-cast patch` commit, you'll see plenty of `exists(Type t | t = foo())` getting simplified.  

If the added type-cast itself would be redundant then I don't emit a type-cast.  
(For the above that would be the case if the type of `foo()` is `Type`, in that case it simplifies to `exists(foo())`).